siphon/simplewebservice/nhc.py

```diff
@@ -0,0 +1,163 @@
+    Read data from the National Hurricane Center Database (NHCD).
+
+    This class reads and then makes dataframes to easier access NHC Data.
+
+    """
+
+    def __init__(self):
+        """
+        Create with member attributes and storm info.
+
+        This initiation creates a file table based on a url for all storms in the
+        NHCD and puts them into a pandas dataframe. This dataframe is then turned
+        into a member atribute '.storm_table'.
+
+        """
+        storm_list_columns = ['Name', 'Basin', 'CycloneNum', 'Year', 'StormType', 'Filename']
+        file_table = pd.read_csv('http://ftp.nhc.noaa.gov/atcf/index/storm_list.txt',
+                                 names=storm_list_columns, header=None, index_col=False,
+                                 usecols=[0, 1, 7, 8, 9, 20])
+        file_table.Filename = file_table.Filename.str.lower()
+        self.storm_table = file_table
+
+    def get_tracks(self, year, filename):
+        """
+        Make url and pulls track data for a given storm.
+
+        The Url is made by using both the year and the filename. This function will then
+        read the data and create a data frame for both the forecast and best tracks and
+        compile these data frames into a dictionary. This function returns this dictionary
+        of forecast and best track.
+
+        Parameters
+        ----------
+        self:
+            storing the storm dictionary as member attrubute of NHCD
+        year: int
+            year of the storm incident
+        filename: str
+            unique filename of the storm which is used for indexing purposes and id
+            in the NHCD. The first character is defaulted as space in NHCD so it is clipped
+            when being used.
+
+        Returns
+        -------
+        unique_models: list
+            all the models that have run forecasts for this storm throughout its life
+
+        """
+        today = datetime.today()
+        current_year = today.year
+        data_dictionary = {}
+        # Current year data is stored in a different location
+        if year == str(current_year):
+            unformatted_forecast_url = 'http://ftp.nhc.noaa.gov/atcf/aid_public/a{}.dat.gz'
+            urlf = unformatted_forecast_url.format(filename[1:])
+            unformatted_best_url = 'http://ftp.nhc.noaa.gov/atcf/btk/b{}.dat'
+            urlb = unformatted_best_url.format(filename[1:])
+        else:
+            unformatted_forecast_url = 'http://ftp.nhc.noaa.gov/atcf/archive/{}/a{}.dat.gz'
+            urlf = unformatted_forecast_url.format(year, filename[1:])
+            unformatted_best_url = 'http://ftp.nhc.noaa.gov/atcf/archive/{}/b{}.dat.gz'
+            urlb = unformatted_best_url.format(year, filename[1:])
+
+        url_links = [urlf, urlb]
+        url_count = 0
+        for url in url_links:
+            # Checking if url is valid, if status_code is 200 then website is active
+            if requests.get(url).status_code == 200:
```

Comment:
There might be cases where you get another value of 2xx code that is actually ok. Rather than checking for 200 exactly, you can use `raise_for_status()` on the response, and if something went wrong (4xx, 5xx), requests will raise a `requests.exceptions.HTTPError` error. See https://3.python-requests.org/user/quickstart/#response-status-codes.

You could replace the whole if block with something like:

~~~python
try:
    response = requests.get(url)
    response .raise_for_status()
    # rest of code after old if statement
    ...
except requests.exceptions.HTTPError as http_error:
    # give helpful message on console
    print("url {} was not valid, select different storm.".format(url))
    # raise the caught error
    raise    
~~~

Reply:
Now I'm wondering...since there is only one direct use of the requests library in your PR, perhaps class `NHCD` should subclass `HTTPEndPoint`, much like `IAStateUpperAir` does. @dopplershift @zbruick? Or is that something we refactor later, given the summer timeline?

Reply:
Yeah I think consistent methods for accessing data whenever possible across the package would be nice. Since I haven't played around with any of this, I'm not sure how much refactoring it would take, but given how `IAStateUpperAir` uses it, I don't think it would take that much time. @AodhanSweeney, chime in if this would take too long - not sure what's left on your to-do list for the summer. If too much, we can refactor down the road.

Reply:
We would still need to check if the URL was valid though, right? `NDBC` still does this with `_check_if_url_valid`, so I don't believe that's handled by `HTTPEndPoint`?

Reply:
I can think of two things to do to address this.

1) replace all `r.status_code == 200` checks with `r.raise_for_status()` (when you have a response obtained from `HTTPEndPoint`)

2) extend `HTTPEndPoint` to have a method for checking if a url is reachable (`is_url_reachable(url)`?) which simply does a head request on the url followed by a `raise_for_status()` (would replace all `_check_if_url_valid` type things). Maybe something like:

~~~python
def is_url_reachable(self, path, params=None):
        """Test if a url is reachable.

        Make a HEAD request, optionally including a parameters, to a path.
        The path of the request is the full URL.
        Parameters
        ----------
        path : str
            The URL to request
        params : DataQuery, optional
            The query to pass when making the request
        Returns
        -------
        resp : requests.Response
            The server's response to the request
        Raises
        ------
        HTTPError
            If the server returns a 4xx or 5xx code

        """
        resp = self._session.head(path, params=params)
        resp.raise_for_status()

        return true
~~~

Since `get` requests follow redirect by default, we should probably make sure the `head` call includes the parameter `allow_redirects=True` (`head` does not follow redirects by default), since the url would ultimately be reachable by a `get` request.

How does that sound?